assertk/src/commonMain/kotlin/assertk/assert.kt

```diff
@@ -190,3 +190,17 @@ inline fun assertFailure(f: () -> Unit): Assert<Throwable> {
     }
     fail("expected failure but lambda completed successfully")
 }
+
+/**
+ * Asserts that the given block will throw an exception with expected type rather than complete successfully.
+ */
+inline fun <reified T: Throwable> assertFailureWith(f: () -> Unit): Assert<T> {
+    @Suppress("TooGenericExceptionCaught", "InstanceOfCheckForException") // Intentionally capturing all exceptions.
+    try {
+        f()
+    } catch (t: Throwable) {
+        if (t !is T) fail("expected failure to be type of ${T::class}")
```

Comment:
I would either:
- Rethrow exceptions of the wrong type allowing the test to fail normally
- Use `assertFailure(f).isInstanceOf(T::class)` as the whole implementation here which will give you a better message and include the exception as the cause

Reply:
I choose the second approach.